### ISSUE TITLE
"The implementations of gRPC in Java and Golang differ in terms of the Status.Code."

### ISSUE THREAD

**Issue by Moonergfp** [grpc/grpc-java#10568]
The error handling approach differs between Java and Golang. Assuming the server responds with an error code of 999.

In Golang client implementations:
```go
	import (
	...
		gstatus "google.golang.org/grpc/status"
	)

	const (
		ServiceNotFoundCode = gcodes.Code(999)
	)

	...

	err = dc.Watch(invalidAppName, invalidSrvName, notifyCh) // gRPC invocation
	grpcStatus, ok := gstatus.FromError(err)

	assert.Equal(t, ServiceNotFoundCode, grpcStatus.Code())
```

In the above Golang code, the error code is set to 999 using gcodes.Code(999). The gstatus.FromError function is used to extract the gRPC status from the error. Then, the code can be compared with the expected ServiceNotFoundCode.

In Java client implementations:
```java
	new StreamObserver<RegisterResponse>() {
	...
	public void onError(Throwable t) {
		Status.Code code = Status.fromThrowable(t).getCode();
	}
}
```

In the Java code snippet, the onError method is invoked when an error occurs. The Status.fromThrowable(t) method is used to extract the gRPC status from the throwable. From the Status, the error code can be obtained using the getCode() method. But the int value is convert to UNKOWN rather than 999.

In summary, the code value in Golang is set to 999 directly, while in Java, the error code is obtained from the Status object. Therefore, there is a difference in how the error codes are handled between the two languages.

I want to know why Java implementations use enums instead of integers.

**Comment by Moonergfp** [comment]
So, in order to transmit error information, I must include it in the gRPC message or provide additional details using Google or another error model.

**Comment by ejona86** [comment]
Status codes aren't for user extension. How exactly an implementation handles unknown status codes might have some variation between languages. I don't see anything wrong with the Java behavior. Go might want to consider if they like their current behavior.

@dfawley, FYI.

**Comment by Moonergfp** [comment]
@ejona86 So,to exchange custom business error codes within throwable stack with a Go server, does i have to utilize the Trailer field of the StatusRuntimeException? I know that a recommended approach for achieving this is to use the StatusProto#fromThrowable function from the io.grpc.protobuf package.

**Comment by ejona86** [comment]
Yeah, use trailers. Seems you are aware of the types of approaches mentioned in https://stackoverflow.com/a/48750825/4690866 . It's your choice as to using trailers directly or using StatusProto.

**Comment by ejona86** [comment]
Seems like this is resolved. If not, comment, and it can be reopened.

(https://github.com/grpc/grpc-go/pull/6662 is trying to make this more obvious on the Go side.)
